### ISSUE TITLE
How to get real client IP in Envoy?

### ISSUE THREAD

**Issue by metalLeveller** [envoyproxy/envoy#9127]
I need to get the real client IP address from the request and as I understand, I can get it with the help of [XFF](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto#envoy-api-field-config-filter-network-http-connection-manager-v2-httpconnectionmanager-use-remote-address)

According to [this documentation](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-for), I need to enable use_remote_address (as it is set to false by the default).

This is an example of a config. I ran tests with different values for skip_xff_append and xff_num_trusted_hops variables. The result is the same - I get the IP address of Envoy node at the x-forwarded-for header.

Configs works for [36f39c746eb7d03b762099b206403935b11972d8](https://hub.docker.com/layers/envoyproxy/envoy/36f39c746eb7d03b762099b206403935b11972d8/images/sha256-23052e099eb4795f8c2f3c330370fa8b9159e66362d6159337de4da885b6334d) version.
```
admin:
  access_log_path: /dev/null
  address:
    socket_address: { address: 0.0.0.0, port_value: 9901 }

static_resources:
  listeners:
    - name: listener_0
      address:
        socket_address: { address: 0.0.0.0, port_value: 8080 }
      filter_chains:
        - filters:
            - name: envoy.http_connection_manager
              config:
                codec_type: auto
                stat_prefix: ingress_http
                stream_idle_timeout: 0s
                use_remote_address: true
                route_config:
                  name: local_route
                  virtual_hosts:
                    - name: local_service
                      domains: ["*"]
                      routes:
                        - match: { prefix: "/api/" }
                          route:
                            prefix_rewrite: "/"
                            cluster: api_gateway
                            max_grpc_timeout: 0s
                            retry_policy:
                              retry_on: 5xx
                              num_retries: 5
                              per_try_timeout: 1s
                      cors:
                        allow_origin: ["*"]
                        allow_methods: GET, PUT, DELETE, POST, OPTIONS
                        allow_headers: keep-alive,user-agent,cache-control,content-type,content-transfer-encoding,x-accept-content-transfer-encoding,x-accept-response-streaming,x-user-agent,x-grpc-web,authorization
                        max_age: "1728000"
                        expose_headers: grpc-status,grpc-message
                        enabled: true
                http_filters:
                  - name: envoy.grpc_web
                  - name: envoy.cors
                  - name: envoy.router
  clusters:
    - name: api_gateway
      connect_timeout: 0.25s
      type: logical_dns
      http2_protocol_options: {}
      lb_policy: round_robin
      hosts: [{ socket_address: { address: "api_gateway", port_value: 6565 }}]
```

**Comment by wbpcode** [comment]
Enable"use_remote_address" will add real address of previous hop to end of XFF. If your client connect envoy directly, you can get real client ip by downstreamDirectRemoteAddress. If there are some other proxy between client and envoy, You must ensure that the intermediate proxy records the client's real IP in the XFF header, and you need to set the correct xff_num_trusted_hops.

I suggest that you open the debug log of envoy, and then observe the XFF header in the request from the client to determine the length of the XFF and whether the real client address exists and its hop count.

Hope this helps you.

**Comment by metalLeveller** [comment]
@wbpcode thanks for your suggestion, but I got the next details:

```
[2020-01-06 07:51:24.796][12][debug][router] [source/common/router/router.cc:549] [C0][S7838114239435758122] router decoding headers:
':method', 'POST'
':scheme', 'http'
':path', '/*******'
':authority', 'grpcproxy'
'x-grpc-web', '1'
'origin', 'https://*******'
'user-agent', 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/79.0.3945.79 Chrome/79.0.3945.79 Safari/537.36'
'content-type', 'application/grpc'
'accept', '*/*'
'sec-fetch-site', 'same-origin'
'sec-fetch-mode', 'cors'
'referer', '*******'
'accept-encoding', 'gzip, deflate, br'
'accept-language', 'en-GB,en;q=0.9,ru-UA;q=0.8,ru;q=0.7,en-US;q=0.6,uk;q=0.5'
'cookie', 'G_ENABLED_IDPS=google'
'x-forwarded-for', '**.*.*.**'
'x-forwarded-proto', 'http'
'x-envoy-internal', 'true'
'x-request-id', '*************************************'
'te', 'trailers'
'grpc-accept-encoding', 'identity,deflate,gzip'
'x-envoy-original-path', '/************************'
```
I see only one IP in x-forwarded-for for the same config, but I've updated version of envoy to envoy:v1.12.2

**Comment by wbpcode** [comment]
From the logs you provided, there are only two possibilities. First, the client connects directly to envoy, and there is no intermediate proxy. Second, the intermediate proxy exists, but it does not add the client IP to x-forwarded-for.

If you just want to test function of envoy, you can just create a fake request with fake x-forwarded-for and observe the log and IP you get.

**Comment by mikexu1976** [comment]
Hello,What is the final solution to get real client IP in Envoy?

**Comment by fai555** [comment]
Any update on this?

**Comment by dongtianyi** [comment]
Anyone has suggestions ?

**Comment by MinusRF** [comment]
What if hop count vary? We can use lua processing in that case but any cleaner way?

**Comment by rootsongjc** [comment]
I noticed that many of you have been inquiring about preserving client source IP in Istio. I recently wrote a blog post titled [Maintaining Traffic Transparency: Preserving Client Source IP in Istio](https://tetrate.io/blog/istio-source-ip-transparency/) that addresses this topic in detail. I hope you find it helpful in solving your source IP preservation challenges. Feel free to check it out and let me know if you have any questions or feedback.
